server/src/ingestion.rs

```diff
@@ -0,0 +1,135 @@
+ */
+
+//! Support for data ingestion from an MQTT broker.
+//!
+//! To use the ingestor, run the "ingestor.start()" method. This method first creates the client. Then
+//! it uses the created client to connect to the broker and subscribe to the specified topics. The
+//! connection message stream is then looped over to ingest the messages that are published to the topics.
+
+use std::{process, time::Duration};
+
+use futures::{executor::block_on, stream::StreamExt};
+use paho_mqtt as mqtt;
+use paho_mqtt::{AsyncClient, AsyncReceiver, Message};
+
+use crate::storage::StorageEngine;
+
+/// A single MQTT client that can subscribe to `broker` and ingest messages from `topics`. Note that
+/// after creation, the ingestor needs to be started to ingest messages.
+pub struct Ingestor {
+    /// Server URI for the MQTT broker.
+    broker: &'static str,
+    /// ID that is used to uniquely identify the ingestor as a client to the MQTT broker.
+    client_id: &'static str,
+    /// Specific topics that should be subscribed to. Use "\[*]" to subscribe to all topics.
+    topics: &'static [&'static str],
+    /// The quality of service for each subscribed-to topic. Should be same length as `topics` field.
+    qos: &'static [i32],
+}
+
+impl Ingestor {
+    pub fn new(
+        broker: &'static str,
+        client_id: &'static str,
+        topics: &'static [&'static str],
+        qos: &'static [i32],
+    ) -> Self {
+        Self {
+            broker,
+            client_id,
+            topics,
+            qos,
+        }
+    }
+
+    /// Create a broker client, subscribe to the topics, and start ingesting messages.
+    pub fn start(self, mut storage_engine: StorageEngine) {
+        println!("Creating MQTT broker client.");
+        let mut client = self.create_client();
+
+        if let Err(err) = block_on(async {
+            let mut stream = self.subscribe_to_broker(&mut client).await;
+
+            println!("Waiting for messages...");
+            Self::ingest_messages(&mut stream, &mut client, &mut storage_engine).await;
+
+            Ok::<(), mqtt::Error>(())
+        }) {
+            eprintln!("{}", err);
+        }
+    }
+
+    /// Create a broker client with the ingestors broker URI and client ID.
+    fn create_client(&self) -> AsyncClient {
+        let create_options = mqtt::CreateOptionsBuilder::new()
+            .server_uri(self.broker)
+            .client_id(self.client_id)
+            .finalize();
+
+        mqtt::AsyncClient::new(create_options).unwrap_or_else(|e| {
+            eprintln!("Error creating the client: {:?}", e);
+            process::exit(1);
+        })
+    }
+
+    /// Make the connection to the broker and subscribe to the specified topics.
+    async fn subscribe_to_broker(
+        &self,
+        client: &mut AsyncClient,
+    ) -> AsyncReceiver<Option<Message>> {
+        // Get message stream before connecting since messages can arrive as soon as the connection is made.
```

Comment:
Maybe replace "Get message stream" with "Get the message stream" and add a comment that documents why 25 was chosen for `buffer_sz`.

Reply:
25 was only chosen because that was the default value used in the paho_mqtt example. The documentation or source does not give any information on how to best set this argument so I thought sticking with their default might be best. Should we mention this in the comment?

Reply:
In general I think it is a good idea to mention were magic numbers come from if possible. So I think we should add a comment stating that it is from the paho_mqtt example.

Reply:
Fixed.